fix(server): Listen only to 127.0.0.1 by default

eirslett opened the pull request:
This pull request supersedes #2865, after discussion that one should rely on the `--host` parameter to determine public/private network interface setup.

eirslett committed:
fix(server): Listen only to 127.0.0.1 by default

patak-js commented:
Copying my comment from the prev PR for reference:
I think we should implement the same scheme as sirv: https://github.com/lukeed/sirv/tree/master/packages/sirv-cli#network-access

> For security reasons, sirv-cli does not expose your server to the network by default. This means that your machine, and only your machine, will be able to access the localhost server.
> If, however, your coworker wants to access the server from their computer, or you want to preview your work on a mobile device, you must use the --host flag. Only then will your server be accessible to other devices on the same network.
> Using --host without a value is equivalent to --host 0.0.0.0, which is makes it discoverable publicly. You may customize this by passing a different value – but you probably don't need to!

@eirslett We should add an equivalent paragraph to the docs. And this PR doesn't implement support for `--host` without a value being equivalent to `--host 0.0.0.0`, no?
If the logic in preview.ts and server/index.ts is shared, we may want to take the opportunity to create a helper now that it is getting more complex. In the case of preview, it is not a big security issue but since sirv is doing this for its server, I agree that it would be good to be consistent.

Shinigami92 reviewed (dismissed):
**suggestion (non-blocking):** The only thing I see here is why do we have so huge duplicate code between server and preview? Can this be shared or e.g. re-called from the server impl?

Shinigami92 commented:
Also it seems there are some failing tests 🤔 
https://github.com/vitejs/vite/pull/2977/checks?check_run_id=2337683223#step:10:16

At least on windows 🤔 @patak-js Do you have windows? Can you reproduce this on your machine?

patak-js commented:
@Shinigami92 I can reproduce this on my machine, but I don't understand how this is related to this PR. Head is green.

eirslett commented:
Yes, we should refactor it, my suggestion is that since this is a security fix, we could merge it (after code review) and do the other improvements in a separate PR. The code duplication should definitely be refactored, and supporting `--host` without arguments sounds like possibly a good idea.

My suspicion is that Windows tries to use the public IP in the tests, but by default the server now only listens to 127.0.0.1, which might be the reason why the tests fail?

Shinigami92 commented:
@eirslett I should consider to create a PR/branch on top of `eirslett:bugfix-security-use-host-param-correctly`

rschristian commented:
> And this PR doesn't implement support for `--host` without a value being equivalent to `--host 0.0.0.0`, no?

FWIW, Luke mentioned he only made that the default due to a few complaints about having to type more to expose. You certainly could make the argument that forcing users to be more explicit when exposing could be of benefit.

Just something to consider.

patak-js commented:
Ok, I agree with @rschristian. We can have `--host` default to `--host 0.0.0.0` if there are a lot of users asking for the shortcut, it is safer to avoid it.

patak-js commented:
About the hint message, I think that it could be less intrusive. I think we should copy sirv here:

![image](https://user-images.githubusercontent.com/583075/114848635-2b9e6200-9ddf-11eb-8f98-63f1d94d6663.png)

It could be:
```
    Network:      Use `--host` to expose
```
Instead of add if we don't default to 0.0.0.0 in this PR

patak-js commented:
The failing example in windows is working well when launching the dev server, but not in the jest tests. This may be an issue with `untilUpdated`. Just in case, @eirslett would you merge main in the PR?

antfu commented:
Agree to keep the `Network:` for hinting. My preference:

```
  > Local:    http://localhost:4000/
  > Network:  (use `--host` to expose)
```

Where the `(...)` is colored with gray or dim.

eirslett committed:
fix(server): Adjust CLI output regarding which IPs the server listens to

eirslett committed:
feat(cli): The --host flag can be run without arguments

eirslett commented:
There we go:

- CLI output looks like @antfu suggested
- The `--host` parameter can be used without arguments

🤞 the tests will pass...

patak-js commented:
It is still failing in Windows, but it should be because of our tests and not because of the feature itself. @eirslett maybe we can force to use the `server.host` option for test so we keep testing in the same environment as before this PR? Tests are working in head, so if we do that, this should also pass, no?

patak-js commented:
Some findings about the failing tests:
Previous to this PR, we are passing `undefined` as the host by default ( `httpServer.listen(port, undefined)` )
- This is equivalent to using `httpServer.listen(port, '::')`, I checked that the tests are passing in Windows passing this string.
- If I force manually the use of `httpServer.listen(port, '0.0.0.0')`, the HMR tests in Windows fails.
- Forcing manually the use of `httpServer.listen(port, '127.0.0.1')`, the HMR tests in Windows also fails.

I don't understand why using `'0.0.0.0'` or `'127.0.0.1'` doesn't work for windows. And I don't know if it is a good idea to force the tests to use `'::'` because we wouldn't be testing the default setup after this PR.

Other details:
- I don't see in the PR, how is `host` being defined as `'0.0.0.0'` in case the value is omitted.
- In the PR If the host is not `'127.0.0.1'`, the cli output will not replace `127.0.0.1` with `localhost` properly.

eirslett commented:
> I don't see in the PR, how is host being defined as '0.0.0.0' in case the value is omitted.

I think the host will then be `undefined` which, when passed to node.js/express, means it will listen on `0.0.0.0`.

> In the PR If the host is not '127.0.0.1', the cli output will not replace 127.0.0.1 with localhost properly.

I don't think it matters - the most valuable piece of information is the public IP address, which you can then pass on to coworkers, or your mobile phone, or whatever should access your dev server remotely.

patak-js commented:
> > I don't see in the PR, how is host being defined as '0.0.0.0' in case the value is omitted.
> 
> I think the host will then be `undefined` which, when passed to node.js/express, means it will listen on `0.0.0.0`.

If `host` is undefined, then `hostname` is going to be '127.0.0.1' with this PR, no?
```js
const hostname = options.host || '127.0.0.1'
```

Note: Looks like if the system supports IPv6, passing undefined defaults to `'::'` instead of `'0.0.0.0'` (but sirv [defaults to '0.0.0.0'](https://github.com/lukeed/sirv/blob/ede9189b6c586cd4697e0ffb16671515fdefecde/packages/sirv-cli/index.js#L82))

> > In the PR If the host is not '127.0.0.1', the cli output will not replace 127.0.0.1 with localhost properly.
> 
> I don't think it matters - the most valuable piece of information is the public IP address, which you can then pass on to coworkers, or your mobile phone, or whatever should access your dev server remotely.

Users may still want to use localhost and give the public IP to others at the same time, and this is how Vite works until this moment. Using `--host` after this PR should work the same way as it is now (so `127.0.0.1` should be replaced by `localhost` in the output)

eirslett committed:
feat(cli): Tweak --host parameter: be even more explicit about defaults

eirslett commented:
I pushed a new change now, which makes the logic even more explicit.

Shinigami92 commented:
Hey @eirslett, I just had a quick look over your PR

Some thoughts I had are:

- Your are checking e.g. `hostname === '127.0.0.1'`, would be `'loaclhost'` also a valid value?
- Is it possible to validate the users input that it is a valid IPv4 address?
- Do you think IPv6 addresses could be also a valid input?

IPv6 support (if wanted) could be outsourced to another PR :thinking:

eirslett commented:
> Your are checking e.g. hostname === '127.0.0.1', would be 'loaclhost' also a valid value?

It's not really checking that, it's just setting the default value to 127.0.0.1.

> Is it possible to validate the users input that it is a valid IPv4 address?

I'm sure it possible, but maybe that's a separate feature?

> Do you think IPv6 addresses could be also a valid input?

Absolutely, there's not really anything in the way now for passing an IPv6 address as the `--host` parameter. This PR only sets the defaults. Anyways I think IPv6 is also a separate issue.

eirslett committed:
fix(tests): listen on host 0.0.0.0 when running tests

The security fix for listening only on 127.0.0.1 by default
seems like it broke tests on Windows. Try using 0.0.0.0 for
tests, and see if that fixes the tests.

eirslett committed:
fix(tests): Try listening on :: to see if that fixes the unit tests on Windows

patak-js commented:
@eirslett using `host: '::'` is working. The only failing test is `glob-import`, because it doesn't have a `vite.config.js`.

Given that the current tests suite is already using `'::'`, because of the [default value when passing an undefined host](https://nodejs.org/api/net.html#net_server_listen):
> If `host` is omitted, the server will accept connections on the unspecified IPv6 address (::) when IPv6 is available, or the unspecified IPv4 address (0.0.0.0) otherwise.

I think that we can continue to use the same for CI and the jest run, and push to another PR the modifications for only using localhost for the tests (if that is possible). We will not be testing the restricted setting at this moment.

But before merging, IMO we should modify `host: true` to pass an undefined host value instead of '0.0.0.0', so users have an easy way to choose the default behavior for `server.listen`.

With this change, we could use `host: true` in the tests. But we shouldn't add this to the playgrounds config files. We can add it to the test setup: https://github.com/vitejs/vite/blob/f3992984258ec47802f92f13fd99fc4e1747e1fd/scripts/jestPerTestSetup.ts#L76 only once and it will be applied to all tests that don't have a custom server. 
There are two tests with a custom server: ssr-vue and ssr-react. In case these are failing, `host: true` can be added to:
- https://github.com/vitejs/vite/blob/f3992984258ec47802f92f13fd99fc4e1747e1fd/packages/playground/ssr-vue/server.js#L40
- https://github.com/vitejs/vite/blob/f3992984258ec47802f92f13fd99fc4e1747e1fd/packages/playground/ssr-react/server.js#L35

What do you think?

nihalgonsalves commented on docs/config/index.md:
```diff
@@ -328,8 +328,15 @@ export default async ({ command, mode }) => {
+- **Default:** `'127.0.0.1'`
 
-  Specify server hostname.
+  Specify server hostname, used to specify which IP addresses the server
+  will be available on.
+  In order to open the server for traffic from other machines, you can
+  set this to `'0.0.0.0'`.
+  
+  It is also possible to set this from the CLI using `--host 0.0.0.0`
+  or just `--host`.
```
[nit]

```suggestion
  Specify which IP addresses the server should listen on.
  Set this to `0.0.0.0` to listen on all addresses, including LAN and public addresses.
  
  This can be set via the CLI using `--host 0.0.0.0` or `--host`.
```

nihalgonsalves commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
 
   const options = config.server || {}
-  const hostname = options.host || 'localhost'
+  let hostname: string
+  if (options.host === undefined) {
+    // Use a secure default
+    hostname = '127.0.0.1'
+  } else if (options.host === true) {
+    // The user probably passed --host in the CLI, without arguments
+    hostname = '0.0.0.0'
+  } else {
+    hostname = options.host as string
+  }
```
Since this is reused in both files, let's extract it to a function:

```ts
const getHostname(hostnameOption: unknown): string {
  if (typeof hostnameOption === 'string') {
    return hostnameOption
  }

  // `--host` without parameters
  if (hostnameOption === true) {
    return '0.0.0.0'
  }

  // secure loopback-only default
  return '127.0.0.1'
}
```

nihalgonsalves commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
-          logger.info(`  > ${type} ${chalk.cyan(url)}`)
-        })
-    )
+    if (hostname === '127.0.0.1') {
+      const url = `${protocol}://localhost:${chalk.bold(port)}${base}`
+      logger.info(`  > Local: ${chalk.cyan(url)}`)
+      logger.info(`  > Network: ${chalk.dim('use `--host` to expose')}`)
+    } else {
+      const interfaces = os.networkInterfaces()
+      Object.keys(interfaces).forEach((key) =>
+        (interfaces[key] || [])
+          .filter((details) => details.family === 'IPv4')
+          .map((detail) => {
+            return {
+              type: detail.address.includes('127.0.0.1')
+                ? 'Local:   '
+                : 'Network: ',
+              host: detail.address
+            }
+          })
+          .forEach(({ type, host }) => {
+            const url = `${protocol}://${host}:${chalk.bold(port)}${base}`
+            logger.info(`  > ${type} ${chalk.cyan(url)}`)
+          })
+      )
+    }
```
This can be extracted too

patak-js replied:
Agreed, but I think it is better if we do these refactorings in a future PR.

nihalgonsalves replied:
👍🏼 Okay to leave this one out, but it would be good to extract the added code (previous comment)

eirslett committed:
improvement to docs

Co-authored-by: Nihal Gonsalves <nihal@nihalgonsalves.com>

eirslett committed:
fix(tests): Revert "fix(tests): Try listening on :: to see if that fixes the unit tests on Windows"

This reverts commit 2fde214e1a64db1db44c8f0ab143fefd6c8c5ac7.

eirslett committed:
fix(tests): Revert "fix(tests): listen on host 0.0.0.0 when running tests"

This reverts commit 491d0c0adebb26f37a3b544cf89d3ecd34bbd058.

eirslett committed:
fix(tests): Server config listen to :: by default in all tests

eirslett commented:
Looks like all the tests pass now! I'm not 100 % sure about making `host: true` mean `hostname: undefined`, but it's certainly possible. I have no idea why `::` vs `0.0.0.0` should make a practical difference (or why the tests fail on Windows because of it)

antfu commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
   const options = config.server || {}
-  const hostname = options.host || 'localhost'
+  let hostname: string
+  if (options.host === undefined) {
```
```suggestion
  if (options.host === undefined || options.host === 'localhost') {
```

Should also normalize `localhost` to `127.0.0.1` as well, otherwise the check on L63 will think  `vite --host localhost` as open.

patak-js commented:
> I'm not 100 % sure about making `host: true` mean `hostname: undefined`, but it's certainly possible. I have no idea why `::` vs `0.0.0.0` should make a practical difference (or why the tests fail on Windows because of it)

@eirslett let's do this change, and we can then use `host: true` in the tests instead of `::`.

eirslett committed:
Update packages/vite/src/node/preview.ts

Co-authored-by: Anthony Fu <anthonyfu117@hotmail.com>

eirslett committed:
fix(server): Use hostname undefined instead of 0.0.0.0 or ::

Shinigami92 commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
+      logger.info(`  > Network: ${chalk.dim('use `--host` to expose')}`)
+    } else {
+      const interfaces = os.networkInterfaces()
+      Object.keys(interfaces).forEach((key) =>
+        (interfaces[key] || [])
```
**suggestion:** Seems we only use the value here :thinking: 
Maybe we should use `Object.values` or `Object.entries` instead?

Shinigami92 commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
+      Object.keys(interfaces).forEach((key) =>
+        (interfaces[key] || [])
+          .filter((details) => details.family === 'IPv4')
+          .map((detail) => {
+            return {
+              type: detail.address.includes('127.0.0.1')
+                ? 'Local:   '
+                : 'Network: ',
+              host: detail.address
+            }
+          })
```
**nitpick:** We can simplify this and get rid of `return` by using
```ts
map((detail) => ({
  type: detail.address.includes('127.0.0.1')
    ? 'Local:   '
    : 'Network: ',
  host: detail.address
})
```

Shinigami92 commented on packages/vite/src/node/preview.ts:
```diff
@@ -41,33 +41,48 @@ export async function preview(
+              host: detail.address
+            }
+          })
+          .forEach(({ type, host }) => {
+            const url = `${protocol}://${host}:${chalk.bold(port)}${base}`
```
**question:** Why are we constructing a `url` here and use `.map` for `type` and `host` above?
Feels a bit inconsistent :thinking: 
For what should `forEach` be used and for what `map`?
Maybe we can merge `map` and `forEach` into one `forEach`?

Shinigami92 commented on packages/vite/src/node/server/index.ts:
```diff
@@ -531,8 +531,17 @@ async function startServer(
+  } else if (options.host === true) {
+    // probably passed --host in the CLI, without arguments
+    hostname = undefined // undefined typically means 0.0.0.0 or :: (listen on all IPs)
+  } else {
+    hostname = options.host as string
```
**question:** Is this also the case if `options.host` is set to `false`?
Cause I see that `host` is of type `string | boolean | undefined`

Shinigami92 replied:
@patak-js as I mentioned here, what is with the `options.host` being `false` case?

![image](https://user-images.githubusercontent.com/7195563/117107363-c0aece00-ad81-11eb-88cf-2c30cf6d3348.png)

The `as string` just hides the TypeScript safeness here ⚠️

eirslett commented:
I mean, yes, the code in there should probably be refactored, prettified, extracted, nitpicked and all, but this is first and foremost a security patch, and it's already been 29 days since I opened the first PR. I'm getting a little bit tired of all the bike shedding, sorry. All I wanted to do was to make Vite more secure.

patak-js commented:
Thanks for the PR @eirslett. For later reference, this was not merged before because tests were not passing til yesterday.

evromalarkey commented:
If you use `host: true` the open option returns undefined - `http://undefined:3000/`

Shinigami92 commented:
> If you use `host: true` the open option returns undefined - `http://undefined:3000/`

@evromalarkey Could you open an issue or maybe directly a PR for that?

evromalarkey commented:
Issue here https://github.com/vitejs/vite/issues/3355
